### DISCUSSION TITLE
How to use the Url Cleaner module to replace characters?

### DISCUSSION THREAD

**Discussion by Murilogs1910** [TrianguloY/URLCheck#247] [Q&A] [answered]
TrianguloY mentioned in the [issue 245](https://github.com/TrianguloY/UrlChecker/issues/245) that it was possible to replace URL encoding ("%2F" and other codes like that) for the actual characters they represent ("/" in that case) using the Url Cleaner.

> I also see that you have some automatic patterns to fix the encoded urls. I think that using the ClearUrl catalog (where you can include you own patterns too) may be more powerful, although the syntax is different.

But after reading the [docs that are linked in the app](https://docs.clearurls.xyz/latest/specs/rules/) I got the impression that it works only for removing strings, not adding/replacing them. After testing (although I didn't really know _what_ to test lol), I got the same impression.

So, if anyone knows if it really is possible and I just missed something, I'd appreciate the help.

Code I tested (there's no replacement there because I just didn't know how to add that):
 ``  "custom": {
    "decode": {
      "urlPattern": "^.*$",
      "rawRules": [
        "%20",
        "%21",
        "%22",
        "%23",
        "%24",
        "%25",
        "%26",
        "%27",
        "%29",
        "%2A",
        "%2B",
        "%2C",
        "%2D",
        "%2E",
        "%2F",
        "%3A",
        "%3B",
        "%3C",
        "%3D",
        "%3E",
        "%3F",
        "%40"
      ]
    }
  },``

**Comment by TrianguloY** [comment, marked as answer]
I think I didn't explain it correctly.

What I was trying to say is that the Url Cleaner module is designed to both remove parameters and extract redirections from them.

In your patterns file you had some in the format (can't remember exactly) something like `"google.com/?redirection=(.*)"` -> `"$1"`.
The issue is that the url parameter is encoded, so the pattern output is something like `"google.com%2F123"`, so you had a bunch of `"%2F"` -> `"/"` automatic replacements.

If you use instead the cleaner module, you should be able to add an entry for that under 'redirections', something like (not tested):
`"custom": { "decode": { "urlPattern": "google.com", "redirections": [ "google.com/?redirection=(.*)" ] } },`. This should correctly extract the pattern and decode it, without any extra step.

In any case, if you prefer the patterns module you can still use it of course!

**Reply by Murilogs1910** [reply]
> In your patterns file you had some in the format (can't remember exactly) something like "google.com/?redirection=(.*)" -> "$1".

You never saw my patterns, so I'm guessing you're talking about the one from the author of issue 245. Well, that doesn't matter.

> If you use instead the cleaner module, you should be able to add an entry for that under 'redirections' [...] This should correctly extract the pattern and decode it, without any extra step.

Your code didn't work, but at least I could understand what I was supposed to do. So, I ended up with this code: `"custom": {
    "decode": {
      "urlPattern": ".*",
      "redirections": [
        "^https?:\/\/.*\/.*\\?.*url=(.*)"
      ]
    }
  },`. It correctly extracts the URL from a 'url' parameter (or other parameters ending in url, e.g. 'mortyurl') and decodes it. This should serve as a great base, and tweaking it so it works in other cases should be easy. Again, thanks for your help!

**Reply by TrianguloY** [reply]
> You never saw my patterns, so I'm guessing you're talking about the one from the author of issue 245. Well, that doesn't matter.

Oops, yes, I didn't check the author 😅 sorry!

Glad it helped. I'll probably add a "decode" parameter anyway, seems useful and very easy to implement.
